### PR TITLE
Add wolfSSL_RAND_set_rand_method() and document ENGINE_load_builtin_engines()

### PR DESCRIPTION
**Pull request by kabuobeid**

### CODE REVIEW

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:17790]: This function can just use the incoming const methods pointer directly. At least that's the assumption openssl makes. Then you can avoid a malloc / copy and free.

There is no locking on the use of `rand_methods` and I think there should be for multi-threaded cases. Let me know if you need some tips for adding the mutex's around the use of `rand_methods`

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:17795]: Please use `WOLFSSL_SUCCESS`.

File: wolfssl/ssl.h
Comment (dgarske) [inline wolfssl/ssl.h:644]: Can we have these callbacks use consistent `len` vs `num`. I prefer `len`.

File: wolfssl/ssl.h
Comment (dgarske) [inline wolfssl/ssl.h:643]: Can you include some minor documentation for the use of each argument in these callbacks?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c]: Updated to use the pointer directly, and added a mutex which locks on all writes/reads to rand_methods.

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c]: Fixed.

File: wolfssl/ssl.h
Comment (kabuobeid) [reply wolfssl/ssl.h]: Updated, but I should note the parameter names were taken from our wolfSSL_RAND functions, so the naming is inconsistent there too.  Should I change their parameter names as well?

File: wolfssl/ssl.h
Comment (kabuobeid) [reply wolfssl/ssl.h:645]: Documentation added for all parameters.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:158]: Please wrap this in `#ifndef SINGLE_THREADED` otherwise you will get a build issue with `OPENSSL_EXTRA` and `SINGLE_THREADED`.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:17884]: I believe you can remove the unused variable code on `add`, `len`, `entropy`. Might be others for these functions.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:17870]: Please remove `NO_WOLFSSL_STUB` on these. They are actually implemented now.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:31574]: Can we change the logic on these, so there is just one unlock flow... don't return ret till bottom? Same with other functions.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c:31574]: Good idea, updated.

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c:158]: Are you sure this is necessary?  I see other mutexes defined without this wrapper, and it builds fine for me without it with SINGLE_THREADED + OPENSSL_EXTRA.  It looks like wolfSSL_Mutex is defined as an int in `wc_port.h` and the mutex methods are overridden, is this not sufficient?

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:17866]: If the lock fails you need to return error here and not call unlock... Same for others below.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:17883]: Same here

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:31576]: Same here

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:31892]: Same here

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:31917]: Same here

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:31949]: Same here

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:31947]: I feel like this locking flow can be cleaned up.  Also why are none of the above also checking `initRandMethMutex`?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c:31947]: This method is being called from our wolfCrypt tests, but wolfSSL_init is not called so this mutex was being used uninitialized and causing an error.  It wouldn't hurt to add it to the other functions, I'll add it and clean this up.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c:17849]: Updated so each function will initialize if the mutex isn't already, then proceed only if initializing succeded.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:4960]: This code section is duplicated about six times. Can you please add a static local function for this to reduce code size? Also if you call this new static local and directly return on an error response then you can eliminate all the `if (initRandMethMutex != 0) {` logic on each of these.

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c:4960]: I will move this logic into a function but to be clear, I thought you didn't want any early returns in these functions.  Are you ok with me returning early in this case?

Comment (dgarske) [reply src/ssl.c:4960]: I am very okay with a return in the middle of the function. My suggestion was to avoid having multiple unlock paths.

**Comment by dgarske** [comment]
@kabuobeid ,

Looks like some Windows Visual Studio compiler issues. Otherwise PR looks great:

```
src/ssl.c: In function ‘int GetRandFunction(RandFunction, void**)’:
src/ssl.c:17901:51: error: invalid conversion from ‘int (*)(const void*, int)’ to ‘void*’ [-fpermissive]
                         *func_ptr = rand_methods->seed;
                                     ~~~~~~~~~~~~~~^~~~
src/ssl.c:17904:51: error: invalid conversion from ‘int (*)(unsigned char*, int)’ to ‘void*’ [-fpermissive]
                         *func_ptr = rand_methods->bytes;
                                     ~~~~~~~~~~~~~~^~~~~
src/ssl.c:17907:51: error: invalid conversion from ‘void (*)()’ to ‘void*’ [-fpermissive]
                         *func_ptr = rand_methods->cleanup;
                                     ~~~~~~~~~~~~~~^~~~~~~
src/ssl.c:17910:51: error: invalid conversion from ‘int (*)(const void*, int, double)’ to ‘void*’ [-fpermissive]
                         *func_ptr = rand_methods->add;
                                     ~~~~~~~~~~~~~~^~~
src/ssl.c:17913:51: error: invalid conversion from ‘int (*)(unsigned char*, int)’ to ‘void*’ [-fpermissive]
                         *func_ptr = rand_methods->pseudorand;
                                     ~~~~~~~~~~~~~~^~~~~~~~~~
src/ssl.c:17916:51: error: invalid conversion from ‘int (*)()’ to ‘void*’ [-fpermissive]
                         *func_ptr = rand_methods->status;
```

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:17879]: This was not what I had in mind. This code is not smaller. Now there is a new `enum RandFunction` and function to get a callback with a void*. I only wanted the mutex init code in a static. Sending you a direct note and will provide you some reference code for what I had in mind.

**Review by dgarske — CHANGES_REQUESTED** [review]

**Review by dgarske — APPROVED** [review]
I pushed a few updates. This PR is approved from my POV and assuming Jenkins passes. Todd should get this PR for a final review.

File: src/ssl.c
Comment (kabuobeid) [inline src/ssl.c:17846]: This function is now always returning WOLFSSL_SUCCESS, even if it fails.  We need an else case after this that returns WOLFSSL_FAILURE, it only succeeds if gRandMethods is set.  Ideally it would report if initializing or locking the mutex failed, but up to you if it's worth the extra lines.

File: src/ssl.c
Comment (kabuobeid) [inline src/ssl.c:17848]: I think it makes more sense to return WOLFSSL_FAILURE if this function is not enabled.

File: src/ssl.c
Comment (kabuobeid) [inline src/ssl.c:17862]: This should be returning WOLFSSL_FAILURE if we failed to determine if there was a random function override due to mutex issues.

File: src/ssl.c
Comment (kabuobeid) [inline src/ssl.c:31577]: If we call this function, we should return early instead of calling wolfSSL_RAND_Init.  These functions are supposed to completely override the RAND functions if they are set.

File: src/ssl.c
Comment (kabuobeid) [inline src/ssl.c]: Why remove this?  The mutex needs to be freed or it will leak memory.  I also think it makes sense to set gRandMethods to NULL on cleanup, since the user is most likely cleaning up their RAND_METHODS struct as well when this is called.

Comment (dgarske) [reply src/ssl.c]: It is done in wolfSSL_RAND_Cleanup()

File: src/ssl.c
Comment (dgarske) [reply src/ssl.c:31577]: Yes but we do not enforce setting all of them and we need to make sure the global rng is initialized. Feel free to improve.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:13248]: Please move this cleanup into `wolfSSL_RAND_Cleanup`.

File: wolfssl/ssl.h
Comment (dgarske) [inline wolfssl/ssl.h:1411]: I think we should revert this change, even though it does not match openssl compatibility because the ABI tracker will flag an issue here. Its fine for an int to return where caller thinks its void, but not for existing customers who might expect a return code here.

File: src/ssl.c
Comment (dgarske) [inline src/ssl.c:4938]: Will you confirm that building `./configure --enable-opensslextra=x509small --disable-rsa` builds correctly with this logic around the new `HAVE_GLOBAL_RNG`?

**Review by dgarske — CHANGES_REQUESTED** [review]

File: src/ssl.c
Comment (kabuobeid) [reply src/ssl.c:4938]: Just tried building with those options and had no errors/warnings, and all tests passed.

**Review by dgarske — APPROVED** [review]

**Review by toddouska — APPROVED** [review]

**Pull request merged**
